### PR TITLE
embulk-output-oracle

### PR DESCRIPTION
**Pull request by hito4t**
This pull request depends on https://github.com/embulk/embulk-output-jdbc/pull/11

### CODE REVIEW

File: embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        OraclePluginTask oracleTask = (OraclePluginTask) task;
+        
+        if (oracleTask.getDriverPath().isPresent()) {
+            loadDriverJar(oracleTask.getDriverPath().get());
+        }
+
+        String url;
+        if (oracleTask.getUrl().isPresent()) {
+            url = oracleTask.getUrl().get();
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java]: How about adding validation here like this?:

```
            if (oracleTask.getHost().isPresent() || oracleTask.getPort().isPresent() || oracleTask.getDatabase().isPresent()) {
                throw new IllegalArgumentException("'host', 'port' and 'database' parameters are invalid if 'url' parameter is set.");
            }
```

**Comment by hito4t** [comment]
Thank you, I think it's better too.
But getPort() is int and doesn't have isPresent().
And even if change getPort() to Optional<Integer>, isPresent() will return always true because default value "1521" exists.

File: embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+                StringBuilder sb = new StringBuilder();
+                sb.append(tablePrefix);
+                for (int j = tablePrefix.length(); j < MAX_TABLE_NAME_LENGTH - s.length(); j++) {
+                    sb.append("0");
+                }
+                sb.append(s);
+
+                String table = sb.toString();
+                if (!connection.tableExists(table)) {
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java]: TODO: this fallback code should be at the parent class.

File: embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                    return table;
+                }
+                
+                if (i == Integer.MAX_VALUE) {
+                	break;
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java]: this looks risky. we can throw an exception.

File: embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                    sb.append("0");
+                }
+                sb.append(s);
+
+                String table = sb.toString();
+                if (!connection.tableExists(table)) {
+                    return table;
+                }
+                
+                if (i == Integer.MAX_VALUE) {
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java]: we can add this condition to the condition part of `for` statement.

File: embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            for (int i = 0; ; i++) {
+                String s = Integer.toString(i);
+                if (tablePrefix.length() + s.length() > MAX_TABLE_NAME_LENGTH) {
+                    break; 
+                }
+
+                StringBuilder sb = new StringBuilder();
+                sb.append(tablePrefix);
+                for (int j = tablePrefix.length(); j < MAX_TABLE_NAME_LENGTH - s.length(); j++) {
+                    sb.append("0");
```
Comment (frsyuki) [inline embulk-output-oracle/src/main/java/org/embulk/output/OracleOutputPlugin.java]: why "0"?

**Comment by frsyuki** [comment]
I created a branch named "oracle3". Can you take a look?:
https://github.com/embulk/embulk-output-jdbc/compare/oracle3

**Comment by hito4t** [comment]
Thank you for your proposal.

If a table name is "TEST123456789AAA", the swap table name is "TEST123456789_<unique id>_bl_tmp".
And if a table name is "TEST123456789BBB", the swap table name is also "TEST123456789_<unique id>_bl_tmp".

I think we cannot avoid this problem (different tables have same swap table name prefix) easily.
So I think we don't need to check if the table name exists (line 140).

And, do we need to check if the generated swap table name exists?

**Comment by frsyuki** [comment]
Good catch. That's right. Simply adding '_' doesn't solve the problem.
To solve the problem, an idea is to add the first (second, third, ... for each loop) character of `md5(original table name)` instead of '_'. It's TODO. I'll remove the checking code for now.

**Comment by frsyuki** [comment]
Merged. Thank you!

**Pull request merged**